File: lang-runners/javascript/stdlib.js
// "use strict";

function checkInstance(val, cls, funcName) {
  if (!(val instanceof cls)) {
    throw new TypeError(
      `${funcName} argument must be an instance of ${cls.name}`,
    )
  }
}

function checkType(val, type, funcName) {
  if (typeof val !== type) {
    throw new TypeError(`${funcName} argument must be of type ${type}`)
  }
}

// https://2ality.com/2020/01/enum-pattern.html
// https://github.com/rauschma/enumify/blob/master/ts/src/index.ts
class Enum {
  static closeEnum() {
    const enumValues = []

    for (const [name, staticInstance] of Object.entries(this)) {
      staticInstance.enumKey = name
      staticInstance.ordinal = enumValues.length
      enumValues.push(staticInstance)
    }

    // Important: only add more static properties *after* processing the enum entries
    this.enumValues = enumValues
  }

  static valueOf(str) {
    checkType(str, 'string', 'Enum.valueOf')
    return this.enumValues.find(val => val.enumKey === str)
  }

  // INSTANCE

  toString() {
    return `${this.constructor.name}.${this.enumKey}`
  }

  toJSON() {
    return this.enumKey
  }
}

class Direction extends Enum {
  get opposite() {
    switch (this) {
      case Direction.East:
        return Direction.West
      case Direction.West:
        return Direction.East
      case Direction.North:
        return Direction.South
      case Direction.South:
        return Direction.North
    }
  }

  get toCoords() {
    switch (this) {
      case Direction.East:
        return new Coords(1, 0)
      case Direction.West:
        return new Coords(-1, 0)
      case Direction.North:
        return new Coords(0, -1)
      case Direction.South:
        return new Coords(0, 1)
    }
  }

  get rotateCw() {
    switch (this) {
      case Direction.North:
        return Direction.East
      case Direction.East:
        return Direction.South
      case Direction.South:
        return Direction.West
      case Direction.West:
        return Direction.North
    }
  }

  get rotateCcw() {
    switch (this) {
      case Direction.North:
        return Direction.West
      case Direction.West:
        return Direction.South
      case Direction.South:
        return Direction.East
      case Direction.East:
        return Direction.North
    }
  }
}
Direction.East = new Direction()
Direction.West = new Direction()
Direction.South = new Direction()
Direction.North = new Direction()
Direction.closeEnum()

class Coords {
  constructor(x, y) {
    checkType(x, 'number', 'Coords constructor')
    checkType(y, 'number', 'Coords constructor')
    this.x = x
    this.y = y
  }

  isSpawn() {
    return SPAWN_COORDS_STRINGS.has(this.toString())
  }

  isHill() {
    return HILL_COORDS_STRINGS.has(this.toString())
  }

  toString() {
    return `(${this.x}, ${this.y})`
  }

  distanceTo(other) {
    checkInstance(other, Coords, 'Coords.distanceTo')
    return Math.sqrt((other.x - this.x) ** 2 + (other.y - this.y) ** 2)
  }

  walkingDistanceTo(other) {
    checkInstance(other, Coords, 'Coords.walkingDistanceTo')
    return Math.abs(other.x - this.x) + Math.abs(other.y - this.y)
  }

  directionTo(other) {
    checkInstance(other, Coords, 'Coords.directionTo')
    const diff = this.sub(other)
    const angle = Math.atan2(diff.y, diff.x)
    if (Math.abs(angle) <= Math.PI / 4) {
      return Direction.West
    } else if (Math.abs(angle + Math.PI / 2) <= Math.PI / 4) {
      return Direction.South
    } else if (Math.abs(angle - Math.PI / 2) <= Math.PI / 4) {
      return Direction.North
    } else {
      return Direction.East
    }
  }

  add(other) {
    if (other instanceof Coords) {
      return new Coords(this.x + other.x, this.y + other.y)
    } else if (other instanceof Direction) {
      return new Coords(this.x + other.toCoords.x, this.y + other.toCoords.y)
    } else {
      throw new TypeError('Coords.add argument must be an instance of Coords or Direction')
    }
  }

  sub(other) {
    if (other instanceof Coords) {
      return new Coords(this.x - other.x, this.y - other.y)
    } else if (other instanceof Direction) {
      return new Coords(this.x - other.toCoords.x, this.y - other.toCoords.y)
    } else {
      throw new TypeError('Coords.sub argument must be an instance of Coords or Direction')
    }
  }

  mul(n) {
    checkType(n, 'number', 'Coords.mul')
    return new Coords(this.x * n, this.y * n)
  }
}

const SPAWN_COORDS = new Set([new Coords(1, 5), new Coords(1, 6), new Coords(1, 7), new Coords(1, 8), new Coords(1, 9), new Coords(1, 10), new Coords(1, 11), new Coords(1, 12), new Coords(1, 13), new Coords(2, 4), new Coords(2, 14), new Coords(3, 3), new Coords(3, 15), new Coords(4, 2), new Coords(4, 16), new Coords(5, 1), new Coords(5, 17), new Coords(6, 1), new Coords(6, 17), new Coords(7, 1), new Coords(7, 17), new Coords(8, 1), new Coords(8, 17), new Coords(9, 1), new Coords(9, 17), new Coords(10, 1), new Coords(10, 17), new Coords(11, 1), new Coords(11, 17), new Coords(12, 1), new Coords(12, 17), new Coords(13, 1), new Coords(13, 17), new Coords(14, 2), new Coords(14, 16), new Coords(15, 3), new Coords(15, 15), new Coords(16, 4), new Coords(16, 14), new Coords(17, 5), new Coords(17, 6), new Coords(17, 7), new Coords(17, 8), new Coords(17, 9), new Coords(17, 10), new Coords(17, 11), new Coords(17, 12), new Coords(17, 13)])
const SPAWN_COORDS_STRINGS = new Set([...SPAWN_COORDS].map(coords => coords.toString()))

const HILL_COORDS = new Set([new Coords(9, 9), new Coords(8, 9), new Coords(8, 8), new Coords(9, 8), new Coords(10, 8), new Coords(10, 9), new Coords(10, 10), new Coords(9, 10), new Coords(8, 10)])
const HILL_COORDS_STRINGS = new Set([...HILL_COORDS].map(coords => coords.toString()))

class Team extends Enum {
  get opposite() {
    if (this === Team.Red) {
      return Team.Blue
    } else return Team.Red
  }
}
Team.Red = new Team()
Team.Blue = new Team()
Team.closeEnum()

class ObjType extends Enum { }
ObjType.Unit = new ObjType()
ObjType.Terrain = new ObjType()
ObjType.closeEnum()

class Obj {
  constructor(obj) {
    checkType(obj, 'object', 'Obj constructor')
    this.__data = obj
  }

  toString() {
    if (this.objType === ObjType.Unit)
      return `<${this.objType} id=${this.id} coords=${this.coords} ${this.team} health=${this.health}>`
    else
      return `<${this.objType} id=${this.id} coords=${this.coords}>`
  }

  get coords() {
    return new Coords(...this.__data.coords)
  }

  get id() {
    return this.__data.id
  }

  get objType() {
    return ObjType.valueOf(this.__data.obj_type)
  }

  get team() {
    if (this.objType === ObjType.Unit) {
      return Team.valueOf(this.__data.team)
    }
  }

  get health() {
    if (this.objType === ObjType.Unit) {
      return this.__data.health
    }
  }
}

class State {
  constructor(state) {
    checkType(state, 'object', 'State constructor')
    this.__data = state
  }

  get turn() {
    return this.__data.turn
  }

  get ourTeam() {
    return Team.valueOf(this.__data.team)
  }

  get otherTeam() {
    return this.ourTeam.opposite
  }

  objById(id) {
    checkType(id, 'string', 'State.objById')
    const obj = this.__data.objs[id]
    if (obj) return new Obj(obj)
  }

  idsByTeam(team) {
    checkInstance(team, Team, 'State.idsByTeam')
    return this.__data.teams[team.enumKey]
  }

  objsByTeam(team) {
    checkInstance(team, Team, 'State.objsByTeam')
    return this.idsByTeam(team).map(id => this.objById(id))
  }

  idByCoords(coords) {
    checkInstance(coords, Coords, 'State.idByCoords')
    return this.__data.grid[coords.y]?.[coords.x]
  }

  objByCoords(coords) {
    checkInstance(coords, Coords, 'State.objByCoords')
    const id = this.idByCoords(coords)
    if (id) return this.objById(id)
  }
}

class ActionType extends Enum { }
ActionType.Attack = new ActionType()
ActionType.Move = new ActionType()
ActionType.Heal = new ActionType()
ActionType.closeEnum()

class Action {
  constructor(type, direction) {
    checkInstance(type, ActionType, 'Action constructor')
    checkInstance(direction, Direction, 'Action constructor')
    this.type = type
    this.direction = direction
  }

  toString() {
    return `<${this.type} ${this.direction}>`
  }

  static move(direction) {
    checkInstance(direction, Direction, 'Action.move')
    return new Action(ActionType.Move, direction)
  }

  static attack(direction) {
    checkInstance(direction, Direction, 'Action.attack')
    return new Action(ActionType.Attack, direction)
  }

  static heal(direction) {
    checkInstance(direction, Direction, 'Action.heal')
    return new Action(ActionType.Heal, direction)
  }
}


const MAP_SIZE = 19


function __format_err(err, isInitError = false) {
  let lineno = null
  if (err) {
    if (err.lineNumber) {
      lineno = err.lineNumber
    } else if (Array.isArray(err.traceback)) {
      for (const entry of err.traceback) {
        if (
          entry &&
          entry.fileName === '<robot>' &&
          Number.isInteger(entry.lineNumber)
        ) {
          lineno = entry.lineNumber
        }
      }
    }
  }
  const e = {
    loc:
      lineno == null
        ? null
        : {
          start: [lineno, null],
          end: null,
        },
    summary: String(err),
    details: (err && err.stack) || null,
  }
  return { Err: isInitError ? { InitError: e } : e }
}

function __main(stateData) {
  function __validateFunction(name, f, argcount, mandatory) {
    if (typeof f !== 'function') {
      if (mandatory) {
        throw new TypeError(`You must define a '${name}' function`)
      }
    } else {
      if (f.length !== argcount) {
        const err = new TypeError(
          `Your ${name} function must accept ${argcount} arguments`,
        )
        err.lineNumber = f && f.lineNumber
        err.fileName = f && f.fileName
        throw err
      }
    }
  }

  const state = new State(stateData)

  try {
    __validateFunction('robot', globalThis.robot, 2, true)
    __validateFunction('initTurn', globalThis.initTurn, 1, false)
  } catch (e) {
    return __format_err(e, true)
  }

  if (typeof globalThis.initTurn === 'function') {
    try {
      globalThis.initTurn(state)
    } catch (e) {
      return __format_err(e, true)
    }
  }

  const logs = []
  globalThis.console = {
    log(...args) {
      logs.push(args.join(' ') + '\n')
    },
  }
  const robot_actions = {}
  const debug_inspect_tables = {}
  const debug_locate_queries = []
  for (const id of state.idsByTeam(state.ourTeam)) {
    const debug_inspect_table = {}

    class Debug {
      inspect(key, val) {
        checkType(key, 'string', 'Debug.inspect "key"')
        debug_inspect_table[key] = String(val)
      }

      locate(unit) {
        checkInstance(unit, Obj, 'Debug.locate')
        debug_locate_queries.push(unit.id)
      }
    }

    globalThis.debug = new Debug()

    let result
    try {
      const output = globalThis.robot(state, state.objById(id))
      if (output instanceof Action) result = { Ok: output }
      else if (output === null) result = { Ok: null }
      else throw new TypeError('Robot must return an Action or null')
    } catch (e) {
      result = __format_err(e)
    }
    robot_actions[id] = result
    if (Object.keys(debug_inspect_table).length) debug_inspect_tables[id] = debug_inspect_table
  }
  return { Ok: { robot_actions, logs, debug_inspect_tables, debug_locate_queries } }
}

File: lang-runners/python/stdlib.py
#!/usr/bin/env python
import enum
import math
from typing import List, Optional, Any


def check_instance(val, cls, func_name):
    if not isinstance(val, cls):
        raise TypeError(f"{func_name} argument must be an instance of {cls.__name__}")


class Direction(enum.Enum):
    North = "North"
    South = "South"
    East = "East"
    West = "West"

    @property
    def opposite(self) -> "Direction":
        return {
            Direction.East: Direction.West,
            Direction.West: Direction.East,
            Direction.South: Direction.North,
            Direction.North: Direction.South,
        }[self]

    @property
    def to_coords(self) -> "Coords":
        return {
            Direction.East: Coords(1, 0),
            Direction.West: Coords(-1, 0),
            Direction.South: Coords(0, 1),
            Direction.North: Coords(0, -1),
        }[self]


class ActionType(enum.Enum):
    Attack = "Attack"
    Move = "Move"


class Coords(tuple):
    def __new__(cls, x: int, y: int) -> "Coords":
        check_instance(x, int, "Coords.__new__")
        check_instance(y, int, "Coords.__new__")
        self = super().__new__(cls, [x, y])
        return self

    def __repr__(self) -> str:
        return f"({self.x}, {self.y})"

    @property
    def x(self) -> int:
        return self[0]

    @property
    def y(self) -> int:
        return self[1]

    def distance_to(self, other: "Coords") -> float:
        check_instance(other, Coords, "Coords.distance_to")
        return math.sqrt((other.x - self.x) ** 2 + (other.y - self.y) ** 2)

    def walking_distance_to(self, other: "Coords") -> int:
        check_instance(other, Coords, "Coords.walking_distance_to")
        return abs(other.x - self.x) + abs(other.y - self.y)

    def coords_around(self) -> List[Direction]:
        return [self + direction for direction in Direction]

    def direction_to(self, other: "Coords") -> Direction:
        check_instance(other, Coords, "Coords.direction_to")
        diff = self - other
        angle = math.atan2(diff.y, diff.x)
        if abs(angle) <= math.pi / 4:
            return Direction.West
        elif abs(angle + math.pi / 2) <= math.pi / 4:
            return Direction.South
        elif abs(angle - math.pi / 2) <= math.pi / 4:
            return Direction.North
        else:
            return Direction.East

    def __add__(self, other: "Coords") -> "Coords":
        check_instance(other, Coords, "Coords.__add__")
        return Coords(self.x + other.x, self.y + other.y)

    def __sub__(self, other: "Coords") -> "Coords":
        check_instance(other, Coords, "Coords.__sub__")
        return Coords(self.x - other.x, self.y - other.y)

    def __mul__(self, n: int) -> "Coords":
        check_instance(n, int, "Coords.__mul__")
        return Coords(self.x * n, self.y * n)


class Team(enum.Enum):
    Red = "Red"
    Blue = "Blue"

    @property
    def opposite(self) -> "Team":
        if self == Team.Red:
            return Team.Blue
        else:
            return Team.Red


class ObjType(enum.Enum):
    Unit = "Unit"
    Terrain = "Terrain"


class Obj:
    def __init__(self, obj: dict) -> None:
        check_instance(obj, dict, "Coords.__init__")
        self.__data = obj

    @property
    def coords(self) -> Coords:
        return Coords(*self.__data["coords"])

    @property
    def id(self) -> str:
        return self.__data["id"]

    @property
    def obj_type(self) -> ObjType:
        return ObjType(self.__data["obj_type"])

    @property
    def team(self) -> Team:
        if self.obj_type == ObjType.Unit:
            return Team(self.__data["team"])

    @property
    def health(self) -> int:
        if self.obj_type == ObjType.Unit:
            return self.__data["health"]


class State:
    def __init__(self, state: dict) -> None:
        check_instance(state, dict, "State.__init__")
        self.__data = state

    @property
    def turn(self) -> int:
        return self.__data["turn"]

    @property
    def our_team(self) -> Team:
        return Team(self.__data["team"])

    @property
    def other_team(self) -> Team:
        return self.our_team.opposite

    def obj_by_id(self, id: str) -> Optional[Obj]:
        check_instance(id, str, 'State.obj_by_id')
        try:
            return Obj(self.__data["objs"][id])
        except KeyError:
            return None

    def ids_by_team(self, team: Team) -> List[str]:
        check_instance(team, Team, 'State.check_instance')
        return self.__data["teams"][team.value]

    def objs_by_team(self, team: Team) -> List[Obj]:
        check_instance(team, Team, 'State.objs_by_team')
        return [self.obj_by_id(id) for id in self.ids_by_team(team)]

    def id_by_coords(self, coords: Coords) -> Optional[str]:
        check_instance(coords, Coords, 'State.id_by_coords')
        try:
            return self.__data["grid"][coords.x][coords.y]
        except IndexError:
            return None

    def obj_by_coords(self, coords: Coords) -> Optional[Obj]:
        check_instance(coords, Coords, 'State.obj_by_coords')
        id = self.id_by_coords(coords)
        if id:
            return self.obj_by_id(id)


class Action:
    def __init__(self, type: ActionType, direction: Direction) -> None:
        check_instance(type, ActionType, 'Action.__init__')
        check_instance(direction, Direction, 'Action.__init__')
        self.type = type
        self.direction = direction

    def __repr__(self) -> str:
        return f"<Action: {self.type} {self.direction}>"

    @staticmethod
    def move(direction: Direction) -> "Action":
        check_instance(direction, Direction, 'Action.move')
        return Action(ActionType.Move, direction)

    @staticmethod
    def attack(direction: Direction) -> "Action":
        check_instance(direction, Direction, 'Action.attack')
        return Action(ActionType.Attack, direction)


def __format_err(exc):
    loc = None
    tb = exc.__traceback__
    while tb:
        if tb.tb_frame.f_code.co_filename == "<robot>":
            loc = {
                "start": (tb.tb_lineno, None),
                "end": None,
            }
        tb = tb.tb_next
    import traceback
    tb_lines = list(traceback.TracebackException.from_exception(exc).format())
    # from docs: "The message indicating which exception occurred is always the last string in the output."
    return {
        "summary": tb_lines.pop().strip(),
        "details": "".join(tb_lines),
        "loc": loc,
    }


def __main(state, scope=globals()):
    def __validate_function(name, argcount, mandatory):
        f = scope.get(name)
        if not callable(f):
            if mandatory:
                raise TypeError(f"You must define a '{name}' function")
        else:
            if f.__code__.co_argcount != argcount:
                raise TypeError(
                    f"Your {name} function must accept {argcount} arguments"
                )
        return f

    import sys, io

    had_stdout, old_stdout = (
        (True, sys.stdout) if hasattr(sys, "stdout") else (False, None)
    )
    logbuf = sys.stdout = io.StringIO()

    state = State(state)
    try:
        robot = __validate_function("robot", 2, True)
        init_turn = __validate_function("init_turn", 1, False)
    except Exception as e:
        return {"Err": {"InitError": __format_err(e)}}

    if callable(init_turn):
        init_turn(state)

    robot_actions = {}
    debug_tables = {}
    debug_inspections = []

    for id in state.ids_by_team(state.our_team):
        global debug
        debug_table = {}

        class Debug:
            def log(self, key: str, val: Any) -> None:
                check_instance(key, str, "Debug.log 'key'")
                debug_table[key] = str(val)

            def inspect(self, unit: Obj) -> None:
                check_instance(unit, Obj, "Debug.inspect")
                debug_inspections.append(unit.id)

        debug = Debug()

        try:
            action = robot(state, state.obj_by_id(id))
            if isinstance(action, Action):
                result = {
                    "Ok": {"type": action.type.value, "direction": action.direction.value}
                }
            elif action is None:
                result = None
            else:
                raise TypeError("Robot must return an Action or None")
        except Exception as e:
            result = {"Err": __format_err(e)}

        if result is not None:
            robot_actions[id] = result
        debug_tables[id] = debug_table

    if had_stdout:
        sys.stdout = old_stdout
    else:
        del sys.stdout

    logbuf.seek(0)
    logs = logbuf.readlines()
    logbuf.close()

    return {
        "Ok": {
            "robot_actions": robot_actions,
            "logs": logs,
            "debug_tables": debug_tables,
            "debug_inspections": debug_inspections
        }
    }


del enum

if __name__ == "__main__":
    __builtins__.__dict__.update(globals())
    import sys, json, runpy

    module = sys.argv[1]
    module = runpy.run_path(module)
    print('__rr_init:{"Ok":null}', flush=True)
    for inp in sys.stdin:
        inp = json.loads(inp)
        output = __main(inp, scope=module)
        sys.stdout.write("__rr_output:")
        json.dump(output, sys.stdout)
        print(flush=True)
